flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/TwoInputSelectable.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.api.operators;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+/**
+ * Interface for the two-input operator that can select the input to get the next
+ * {@link org.apache.flink.streaming.runtime.streamrecord.StreamRecord}.
+ */
+@PublicEvolving
+public interface TwoInputSelectable {
+
+	/**
+	 * Returns the input selection that wants to get the next record.
+	 */
+	InputIdentifier nextSelection();
```

Comment:
This interface would also work to select 1..n inputs, so why bind it to exaclty two? Furthermore, while I like the OO approach in `InputIdentifier`, I wonder if this isn't a very hot code path and if an `int` or `long` would not be better in that case, especially when it represents a bitmask and not an identifier.

Reply:
It will be executed once per record. TBH, I'm not entirely sure if this is a premature optimisation or not. We can either:

1. Go with this wrapper class initially, and maybe replace it after benchmarking the runtime code. For me goal of this effort should be to `StreamTwoInputSelectableProcessor` have the same performance as regular `StreamTwoInputProcessor` assuming constant/static input selection (so that we can implement one using the other's code). 
2. Wait with the merging of this PR until the runtime code is ready and we have better picture of which way we should go.

So far in previous efforts we were merging API changes along the runtime parts in one chunk, so I think we should follow the 2nd option in this case as well.

Reply:
**object VS int or long**
For reasons, see the suggestion by StephanEwen: http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/DISCUSS-Enhance-Operator-API-to-Support-Dynamically-Selective-Reading-and-EndOfInput-Event-tt26753.html#a26816

**why bind it to exaclty two?**
Currently there is only `TwoInputStreamOperator` interface, I think it is better to bind it to exaclty two. In the future, if `NInputStreamOperator` is introduced, an `InputSelectable` interface can be introduced correspondingly.

Reply:
Please move to https://github.com/apache/flink/pull/7959#issuecomment-474219792 to see the runtime code.